File: TI2.1  SOAP_JAVA_EUREKA_BANK_GR08/01.SERVIDOR/ws_eureka_bank_soap_java/src/main/java/ec/edu/monster/servicio/CuentaService.java
package ec.edu.monster.servicio;

import ec.edu.monster.Controlador.CuentaDAO;
import ec.edu.monster.Controlador.MovimientoDAO;
import ec.edu.monster.modelo.CuentaModel;
import ec.edu.monster.modelo.MovimientoModel;
import java.sql.SQLException;
import java.time.LocalDate;
import java.util.List;

public class CuentaService {

    private MovimientoDAO movimientoDAO = new MovimientoDAO();
    private CuentaDAO cuentaDAO = new CuentaDAO();

    // Método para actualizar el saldo de la cuenta y registrar el movimiento
    public boolean actualizarSaldoYRegistrarMovimiento(String codigoCuenta, String valorMovimiento, String tipo, String cuentaDest) {
        try {
            double importe = Double.parseDouble(valorMovimiento);
            String codTipo = "";

            // Validate the operation type
            if (tipo.equalsIgnoreCase("RET")) {  // Retiro
                codTipo = "004";

                CuentaModel cuenta = cuentaDAO.obtenerCuentaPorCodigo(codigoCuenta);
                if (cuenta.getDecCuenSaldo() < importe) {
                    return false;
                }

                if (!cuentaDAO.actualizarSaldoCuenta(codigoCuenta, String.valueOf(-importe))) {
                    return false;  // Return false if balance couldn't be updated
                }
            } else if (tipo.equalsIgnoreCase("DEP")) {
                codTipo = "003";
                if (!cuentaDAO.actualizarSaldoCuenta(codigoCuenta, valorMovimiento)) {
                    return false;  // Return false if balance couldn't be updated
                }
            } else if (tipo.equalsIgnoreCase("TRA")) {  // Transferencia
                // Check that the destination account is provided
                codTipo = "009";
                if (cuentaDest == null || cuentaDest.isEmpty()) {
                    throw new IllegalArgumentException("Cuenta destino es obligatoria para transferencias.");
                }

                CuentaModel cuentaOrigen = cuentaDAO.obtenerCuentaPorCodigo(codigoCuenta);
                if (cuentaOrigen.getDecCuenSaldo() < importe) {
                    return false;
                }

                // Deduct from source account
                if (!cuentaDAO.actualizarSaldoCuenta(codigoCuenta, String.valueOf(-importe))) {
                    return false;  // Return false if balance couldn't be updated
                }

                // Add to destination account
                if (!cuentaDAO.actualizarSaldoCuenta(cuentaDest, valorMovimiento)) {
                    // Rollback source account update if destination update fails
                    cuentaDAO.actualizarSaldoCuenta(codigoCuenta, valorMovimiento);
                    return false;
                }
            } else {
                throw new IllegalArgumentException("Tipo de movimiento no soportado: " + tipo);
            }

            // Register the movement
            int numeroMovimiento = obtenerSiguienteNumeroMovimiento(codigoCuenta);

            MovimientoModel movimiento = new MovimientoModel();
            movimiento.setCodigoCuenta(codigoCuenta);
            movimiento.setNumeroMovimiento(numeroMovimiento);
            movimiento.setFechaMovimiento(LocalDate.now().toString());
            movimiento.setCodigoEmpleado("0001");  // Update as needed
            movimiento.setCodigoTipoMovimiento(codTipo);
            movimiento.setImporteMovimiento(importe);
            
            // Para transferencias, guardar la cuenta destino en cuentaReferencia
            if (tipo.equalsIgnoreCase("TRA")) {
                movimiento.setCuentaReferencia(cuentaDest);
            }

            movimientoDAO.registrarMovimiento(movimiento);

            // Register the movement for the destination account in case of transfer
            if (tipo.equalsIgnoreCase("TRA")) {
                int numeroMovimientoDest = obtenerSiguienteNumeroMovimiento(cuentaDest);

                MovimientoModel movimientoDest = new MovimientoModel();
                movimientoDest.setCodigoCuenta(cuentaDest);
                movimientoDest.setNumeroMovimiento(numeroMovimientoDest);
                movimientoDest.setFechaMovimiento(LocalDate.now().toString());
                movimientoDest.setCodigoEmpleado("0001");  // Update as needed
                movimientoDest.setCodigoTipoMovimiento("008");
                movimientoDest.setImporteMovimiento(importe);
                // Para transferencias entrantes, guardar la cuenta origen en cuentaReferencia
                movimientoDest.setCuentaReferencia(codigoCuenta);

                movimientoDAO.registrarMovimiento(movimientoDest);
            }

            return true;
        } catch (SQLException | IllegalArgumentException e) {
            e.printStackTrace();
            return false;
        }
    }

    // Método para obtener el siguiente número de movimiento para una cuenta
    private int obtenerSiguienteNumeroMovimiento(String codigoCuenta) throws SQLException {
        List<MovimientoModel> movimientos = movimientoDAO.obtenerMovimientosPorCuenta(codigoCuenta);
        int numeroMovimiento = 1;  // Empezamos con el número de movimiento 1 por defecto

        // Buscamos el último número de movimiento
        for (MovimientoModel movimiento : movimientos) {
            if (movimiento.getNumeroMovimiento() >= numeroMovimiento) {
                numeroMovimiento = movimiento.getNumeroMovimiento() + 1;
            }
        }
        return numeroMovimiento;
    }

    public CuentaModel obtenerCuenta(String codigoCuenta) throws SQLException {
        return cuentaDAO.obtenerCuentaPorCodigo(codigoCuenta);
    }
}

File: Tl2.2 SOAP_DOTNET_EUREKA_BANK_GR08/01.SERVIDOR/ws_eureka_bank_soap_dotnet/ec.edu.monster.service/CuentaService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ws_eureka_bank_soap_dotnet.ec.edu.monster.DAO;
using ws_eureka_bank_soap_dotnet.ec.edu.monster.model;

namespace ws_eureka_bank_soap_dotnet.ec.edu.monster.service
{
    public class CuentaService
    {
        private readonly MovimientoDAO _movimientoDAO = new MovimientoDAO();
        private readonly CuentaDAO _cuentaDAO = new CuentaDAO();

        public bool ActualizarSaldoYRegistrarMovimiento(string codigoCuenta, string valorMovimiento, string tipo, string cuentaDest)
        {
            try
            {
                double importe = double.Parse(valorMovimiento);
                string codTipo = "";

                if (tipo.Equals("RET", StringComparison.OrdinalIgnoreCase))
                {
                    codTipo = "004";
                    var cuenta = _cuentaDAO.ObtenerCuentaPorCodigo(codigoCuenta);
                    if (cuenta == null || cuenta.DecCuenSaldo < importe)
                        return false;

                    if (!_cuentaDAO.ActualizarSaldoCuenta(codigoCuenta, -importe))
                        return false;
                }
                else if (tipo.Equals("DEP", StringComparison.OrdinalIgnoreCase))
                {
                    codTipo = "003";
                    if (!_cuentaDAO.ActualizarSaldoCuenta(codigoCuenta, importe))
                        return false;
                }
                else if (tipo.Equals("TRA", StringComparison.OrdinalIgnoreCase))
                {
                    codTipo = "009";
                    if (string.IsNullOrEmpty(cuentaDest))
                        throw new ArgumentException("La cuenta destino es obligatoria para transferencias.");

                    var cuentaOrigen = _cuentaDAO.ObtenerCuentaPorCodigo(codigoCuenta);
                    if (cuentaOrigen == null || cuentaOrigen.DecCuenSaldo < importe)
                        return false;

                    if (!_cuentaDAO.ActualizarSaldoCuenta(codigoCuenta, -importe))
                        return false;

                    if (!_cuentaDAO.ActualizarSaldoCuenta(cuentaDest, importe))
                    {
                        _cuentaDAO.ActualizarSaldoCuenta(codigoCuenta, importe); // Rollback
                        return false;
                    }
                }
                else
                {
                    throw new ArgumentException($"Tipo de movimiento no soportado: {tipo}");
                }

                int numeroMovimiento = ObtenerSiguienteNumeroMovimiento(codigoCuenta);

                var movimiento = new MovimientoModel
                {
                    CodigoCuenta = codigoCuenta,
                    NumeroMovimiento = numeroMovimiento,
                    FechaMovimiento = DateTime.Now.ToString("yyyy-MM-dd"),
                    CodigoEmpleado = "0001",
                    CodigoTipoMovimiento = codTipo,
                    ImporteMovimiento = importe,
                    CuentaReferencia = cuentaDest
                };

                _movimientoDAO.RegistrarMovimiento(movimiento);

                if (tipo.Equals("TRA", StringComparison.OrdinalIgnoreCase))
                {
                    int numeroMovimientoDest = ObtenerSiguienteNumeroMovimiento(cuentaDest);
                    var movimientoDest = new MovimientoModel
                    {
                        CodigoCuenta = cuentaDest,
                        NumeroMovimiento = numeroMovimientoDest,
                        FechaMovimiento = DateTime.Now.ToString("yyyy-MM-dd"),
                        CodigoEmpleado = "0001",
                        CodigoTipoMovimiento = "008",
                        ImporteMovimiento = importe
                    };
                    _movimientoDAO.RegistrarMovimiento(movimientoDest);
                }

                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
        }

        private int ObtenerSiguienteNumeroMovimiento(string codigoCuenta)
        {
            var movimientos = _movimientoDAO.ObtenerMovimientosPorCuenta(codigoCuenta);
            int numeroMovimiento = 1;

            foreach (var movimiento in movimientos)
            {
                if (movimiento.NumeroMovimiento >= numeroMovimiento)
                    numeroMovimiento = movimiento.NumeroMovimiento + 1;
            }
            return numeroMovimiento;
        }

        public CuentaModel ObtenerCuenta(string codigoCuenta)
        {
            return _cuentaDAO.ObtenerCuentaPorCodigo(codigoCuenta);
        }
    }
}
